From 61a79a52ed6d7a56879e9e14aeadbbb96c2b1dce Mon Sep 17 00:00:00 2001
From: saravananm <saravananm@ami.com>
Date: Tue, 7 Jan 2025 09:59:45 +0530
Subject: [PATCH] Added feature for Certificate replacement

Description:
Return status code 500 if replace HTTP certificate for several times

Tested:
Verified and build taken successfully.

Testcases:

root@AMIOT-a64404d0d99f:~# busctl set-property xyz.openbmc_project.Certs.Manager.Server.Https /xyz/openbmc_project/certs/server/http            s/1 xyz.openbmc_project.Certs.Certificate CertificateType s "xyz.openbmc_project.Certs.Certificate.Type.PEMchain"
root@AMIOT-a64404d0d99f:~# busctl introspect xyz.openbmc_project.Certs.Manager.Server.Https /xyz/openbmc_project/certs/server/https/            1
NAME                                  TYPE      SIGNATURE RESULT/VALUE                             FLAGS
org.freedesktop.DBus.Introspectable   interface -         -                                        -
.Introspect                           method    -         s                                        -
org.freedesktop.DBus.Peer             interface -         -                                        -
.GetMachineId                         method    -         s                                        -
.Ping                                 method    -         -                                        -
org.freedesktop.DBus.Properties       interface -         -                                        -
.Get                                  method    ss        v                                        -
.GetAll                               method    s         a{sv}                                    -
.Set                                  method    ssv       -                                        -
.PropertiesChanged                    signal    sa{sv}as  -                                        -
xyz.openbmc_project.Certs.Certificate interface -         -                                        -
.CertificateString                    property  s         "-----BEGIN CERTIFICATE-----\nMIICVTC... emits-change writable
.CertificateType                      property  s         "xyz.openbmc_project.Certs.Certificat... emits-change writable
.ChainCertString                      property  s         "-----BEGIN CERTIFICATE-----\nMIID1zC... emits-change writable
.Issuer                               property  s         "C=US,O=OpenBMC,CN=AMIOT-a64404d0d99f"   emits-change writable
.KeyUsage                             property  as        2 "KeyEncipherment" "ServerAuthentica... emits-change writable
.Subject                              property  s         "C=US,O=OpenBMC,CN=AMIOT-a64404d0d99f"   emits-change writable
.UefiSignatureOwner                   property  s         ""                                       emits-change writable
.ValidNotAfter                        property  t         2051238585                               emits-change writable
.ValidNotBefore                       property  t         1735878585                               emits-change writable
xyz.openbmc_project.Certs.Replace     interface -         -                                        -
.Replace                              method    s         -                                        -
xyz.openbmc_project.Object.Delete     interface -         -                                        -
.Delete                               method    -         -                                        -
root@AMIOT-a64404d0d99f:~# busctl call xyz.openbmc_project.Certs.Manager.Server.Https /xyz/openbmc_project/certs/server/https/1 xyz.            openbmc_project.Certs.Replace Replace s /home/root/primary_cacert.pem
root@AMIOT-a64404d0d99f:~# busctl introspect xyz.openbmc_project.Certs.Manager.Server.Https /xyz/openbmc_project/certs/server/https/            1
NAME                                  TYPE      SIGNATURE RESULT/VALUE                             FLAGS
org.freedesktop.DBus.Introspectable   interface -         -                                        -
.Introspect                           method    -         s                                        -
org.freedesktop.DBus.Peer             interface -         -                                        -
.GetMachineId                         method    -         s                                        -
.Ping                                 method    -         -                                        -
org.freedesktop.DBus.Properties       interface -         -                                        -
.Get                                  method    ss        v                                        -
.GetAll                               method    s         a{sv}                                    -
.Set                                  method    ssv       -                                        -
.PropertiesChanged                    signal    sa{sv}as  -                                        -
xyz.openbmc_project.Certs.Certificate interface -         -                                        -
.CertificateString                    property  s         "-----BEGIN CERTIFICATE-----\nMIIEIjC... emits-change writable
.CertificateType                      property  s         "xyz.openbmc_project.Certs.Certificat... emits-change writable
.ChainCertString                      property  s         "-----BEGIN CERTIFICATE-----\nMIIF7zC... emits-change writable
.Issuer                               property  s         "C=US,ST=Georgia,L=Norcross,O=America... emits-change writable
.KeyUsage                             property  as        1 "KeyEncipherment"                      emits-change writable
.Subject                              property  s         "C=US,ST=Georgia,L=Duluth,O=AMI,OU=ME... emits-change writable
.UefiSignatureOwner                   property  s         ""                                       emits-change writable
.ValidNotAfter                        property  t         2051604121                               emits-change writable
.ValidNotBefore                       property  t         1736244121                               emits-change writable
xyz.openbmc_project.Certs.Replace     interface -         -                                        -
.Replace                              method    s         -                                        -
xyz.openbmc_project.Object.Delete     interface -         -                                        -
.Delete                               method    -         -                                        -
root@AMIOT-6eeb1f95f299:~#

Working as expected.

Signed-off-by: saravananm <saravananm@ami.com>
---
 certs_manager.cpp | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/certs_manager.cpp b/certs_manager.cpp
index 380f75b..5177898 100644
--- a/certs_manager.cpp
+++ b/certs_manager.cpp
@@ -525,9 +525,15 @@ void Manager::replaceCertificate(Certificate* const certificate,
         chainFile = chainFile / defaultHttpsChainListFileName;
         if (fs::exists(chainFile))
         {
-            elog<NotAllowed>(
-                NotAllowedReason("Certificate Chain already exist"));
-            return;
+            try
+            {
+                fs::remove(chainFile);
+            }
+            catch (const std::exception& e)
+            {
+                lg2::error("Failed to deleteCertificate");
+                elog<InternalFailure>();
+            }
         }
         installChain_Cert(filePath);
     }
-- 
2.34.1

